python/tests/test_approximate_nearest_neighbors.py

```diff
@@ -95,45 +95,136 @@ def test_example(
         assert obj._cuml_params["algo_params"] == algoParams
 
 
+class ANNEvaluator:
+    """
+    obtain exact knn distances and indices
+    """
+
+    def __init__(self, X: np.ndarray, n_neighbors: int, metric: str) -> None:
+        self.X = X
+        self.n_neighbors = n_neighbors
+        self.metric = metric
+        if metric == "inner_product":
+            from cuml import NearestNeighbors as cuNN
+
+            cuml_knn = cuNN(
+                algorithm="brute",
+                n_neighbors=n_neighbors,
+                output_type="numpy",
+                metric=metric,
+            )
+            cuml_knn.fit(X)
+            self.distances_exact, self.indices_exact = cuml_knn.kneighbors(X)
+        else:
+            from sklearn.neighbors import NearestNeighbors as skNN
+
+            sk_knn = skNN(algorithm="brute", n_neighbors=n_neighbors, metric=metric)
+            sk_knn.fit(X)
+            self.distances_exact, self.indices_exact = sk_knn.kneighbors(X)
+
+    def get_distances_exact(self) -> np.ndarray:
+        return self.distances_exact
+
+    def get_indices_exact(self) -> np.ndarray:
+        return self.indices_exact
+
+    def cal_avg_recall(self, indices_ann: np.ndarray) -> float:
+        assert indices_ann.shape == self.indices_exact.shape
+        assert indices_ann.shape == (len(self.X), self.n_neighbors)
+        retrievals = [
+            np.intersect1d(a, b) for a, b in zip(indices_ann, self.indices_exact)
+        ]
+        recalls = np.array([len(nns) / self.n_neighbors for nns in retrievals])
+        return recalls.mean()
+
+    def cal_avg_dist_gap(self, distances_ann: np.ndarray) -> float:
+        assert distances_ann.shape == self.distances_exact.shape
+        assert distances_ann.shape == (len(self.X), self.n_neighbors)
+        gaps = np.abs(distances_ann - self.distances_exact)
+        return gaps.mean()
+
+    def compare_with_cuml_sg(
+        self,
+        algorithm: str,
+        algoParams: Optional[Dict[str, Any]],
+        given_indices: np.ndarray,
+        given_distances: np.ndarray,
+        tolerance: float,
+    ) -> None:
+        # compare with cuml sg ANN on avg_recall and avg_dist_gap
+        from cuml import NearestNeighbors as cuNN
+
+        cuml_ivfflat = cuNN(
+            algorithm=algorithm,
+            algo_params=algoParams,
+            n_neighbors=self.n_neighbors,
+            metric=self.metric,
+        )
+        cuml_ivfflat.fit(self.X)
+        cumlsg_distances, cumlsg_indices = cuml_ivfflat.kneighbors(self.X)
+        if self.metric == "euclidean" or self.metric == "l2":
+            cumlsg_distances **= 2  # square up cuml distances to get l2 distances
+
+        # compare cuml sg with given results
+        avg_recall_cumlann = self.cal_avg_recall(cumlsg_indices)
```

Comment:
Would it be better to instead compare the given indices and distances  directly to what is returned by sg (instead of comparing the respective recalls and diff from the exact case)?   Or is the distributed setting too different from sg for this?

Reply:
Yeah, distributed setting sometimes can be too different from sg. distance gap can be as large as 0.05 (current tolerance is 0.001). 

In fact, same error happens when running SG twice, the distance gap can be large. 

It seems the returned distances are sensitive to the quantization codebook. And codebooks are not the same when running SG multiple times.